crates/proof-of-sql-parser/src/posql_time/timestamp.rs

```diff
@@ -0,0 +1,226 @@
+        let nanoseconds = dt.timestamp_subsec_nanos();
+        let timeunit = if nanoseconds % 1_000 != 0 {
+            PoSQLTimeUnit::Nanosecond
+        } else if nanoseconds % 1_000_000 != 0 {
+            PoSQLTimeUnit::Microsecond
+        } else if nanoseconds % 1_000_000_000 != 0 {
+            PoSQLTimeUnit::Millisecond
+        } else {
+            PoSQLTimeUnit::Second
+        };
+
+        Ok(PoSQLTimestamp {
+            timestamp: dt.with_timezone(&Utc),
+            timeunit,
+            timezone,
+        })
+    }
+
+    /// Attempts to parse a timestamp string into an `PoSQLTimestamp` structure.
+    /// This function supports two primary formats:
+    ///
+    /// **Unix Epoch Time Parsing**:
+    ///    - Since Unix epoch timestamps don't inherently carry timezone information,
+    ///      any Unix time parsed directly from an integer is assumed to be in UTC.
+    ///
+    /// # Examples
+    /// ```
+    /// use chrono::{DateTime, Utc};
+    /// use proof_of_sql_parser::posql_time::{timestamp::PoSQLTimestamp, timezone::PoSQLTimeZone};
+    ///
+    /// // Parsing a Unix epoch timestamp (assumed to be seconds and UTC):
+    /// let unix_time = 1231006505;
+    /// let intermediate_timestamp = PoSQLTimestamp::to_timestamp(unix_time).unwrap();
+    /// assert_eq!(intermediate_timestamp.timezone, PoSQLTimeZone::Utc);
+    /// ```
+    pub fn to_timestamp(epoch: i64) -> Result<Self, PoSQLTimestampError> {
+        match Utc.timestamp_opt(epoch, 0) {
+            LocalResult::Single(timestamp) => Ok(PoSQLTimestamp {
+                timestamp,
+                timeunit: PoSQLTimeUnit::Second,
+                timezone: timezone::PoSQLTimeZone::Utc,
+            }),
+            LocalResult::Ambiguous(_, _) => Err(PoSQLTimestampError::Ambiguous),
+            LocalResult::None => Err(PoSQLTimestampError::LocalTimeDoesNotExist),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_unix_epoch_time_timezone() {
+        let unix_time = 1231006505; // Unix time as string
+        let expected_timezone = timezone::PoSQLTimeZone::Utc; // Unix time should always be UTC
+        let result = PoSQLTimestamp::to_timestamp(unix_time).unwrap();
+        assert_eq!(result.timezone, expected_timezone);
+    }
+
+    #[test]
+    fn test_unix_epoch_timestamp_parsing() {
+        let unix_time = 1231006505; // Example Unix timestamp (seconds since epoch)
+        let expected_datetime = Utc.timestamp_opt(unix_time, 0).unwrap();
+        let expected_unit = PoSQLTimeUnit::Second; // Assuming basic second precision for Unix timestamp
+        let input = unix_time; // Simulate input as string since Unix times are often transmitted as strings
+        let result = PoSQLTimestamp::to_timestamp(input).unwrap();
+
+        assert_eq!(result.timestamp, expected_datetime);
+        assert_eq!(result.timeunit, expected_unit);
+    }
+}
+
+#[cfg(test)]
+mod rfc3339_tests {
```

Comment:
Is there any reason why this needs to be a separate module?

Reply:
the tests were grouped into classes of specific tests to isolate parsing errors that were happening.